src/crimson/os/seastore/cached_extent.h

```diff
@@ -153,6 +153,52 @@ struct trans_spec_view_t {
 };
 
 class ExtentIndex;
+
+using region_list_t = std::list<std::pair<extent_len_t, extent_len_t>>;
+/// manage small chunks of extent
+class BufferSpace {
+public:
+  BufferSpace() = default;
+  BufferSpace(const BufferSpace &other) {
+      for(auto it = other.buffer_map.begin();
+      it != other.buffer_map.end();
+      ++it) {
+        auto listptr = new ceph::bufferlist(*it->second);
+        buffer_map[it->first].reset(listptr);
+      }
+  }
+  BufferSpace(BufferSpace &&other)
+    : buffer_map(std::move(other.buffer_map)) {}
+
+  struct deep_copy_t {};
+  BufferSpace(const BufferSpace &other, deep_copy_t) {
+      for(auto it = other.buffer_map.begin();
+      it != other.buffer_map.end();
+      ++it) {
+        auto other_list = it->second.get();
+        ceph::bufferptr ptr(other_list->c_str(), other_list->length());
+        auto listptr = new ceph::bufferlist();
+        listptr->append(ptr);
+        buffer_map[it->first].reset(listptr);
+      }
+  }
+
+  void add_buffer(extent_len_t offset, ceph::bufferptr&& ptr);
+
+  ceph::bufferlist get_data(extent_len_t offset, extent_len_t length);
+
+  ceph::bufferptr build_ptr(extent_len_t length);
+
+  bool check_buffer(extent_len_t offset, extent_len_t length);
+
+  region_list_t read_buffer(extent_len_t offset, extent_len_t length);
+
+private:
+  /// map extent range onto buffers.
+  /// key: offset in extent -> value:buffer start from offset
+  std::map<extent_len_t, std::unique_ptr<ceph::bufferlist>> buffer_map;
```

Comment:
I think maybe we can directly use `std::map<extent_len_t, ceph::bufferlist>`, in which case we can also use the default copy constructor for `BufferSpace`. `ceph::bufferlist` has a zero-copy copy constructor.